### PR TITLE
Test2

### PR DESCRIPTION
**Pull request by otobin**
Added algorithms and tests for optional guests. Currently all tests pass.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,141 @@
...
+            eventAttendees.retainAll(attendees);
+            // ^^ eventAttendees now only has the intersection of the two lists
+            if (!eventAttendees.isEmpty()) {
+                unavailableTimes.add(event.getWhen());
+            }
+        }
+        return unavailableTimes;
+    }
+
+    // Given an array list of unavailable unavailable TimeRanges ordered by start time that are
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:91]: unavailable repeated twice

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,141 @@
...
+            }
+        }
+
+        // Get the unavailable times for both the required attendees and optional attendees
+        ArrayList<TimeRange> requiredUnavailableTime = getUnavailableTimes(requiredAttendees, events);
+        ArrayList<TimeRange> optionalUnavailableTime = getUnavailableTimes(optionalAttendees, events);
+
+        // If there are no meetings for the required attendees, return the entire day
+        if (requiredUnavailableTime.size() == 0) {
+            return entireDay;
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: In this case don't you want to find available times for optionalAttendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,141 @@
...
+    // can be booked.
+    public Collection <TimeRange> query(Collection <Event> events, MeetingRequest request) {
+        HashSet<String> requiredAttendees = new HashSet<String> (request.getAttendees());
+        HashSet<String> optionalAttendees = new HashSet<String> (request.getOptionalAttendees());
+        long duration = request.getDuration();
+        Collection<TimeRange> entireDay = new ArrayList<TimeRange> (Arrays.asList(TimeRange.WHOLE_DAY));
+        ArrayList<TimeRange> unavailableTimes = new ArrayList<TimeRange> ();
+
+        // If the duration is over a day, then there is no time available
+        if (duration> TimeRange.WHOLE_DAY.duration()) {
```
Comment (kyracq) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: nit: missing space after "duration"

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,141 @@
...
+        ArrayList<TimeRange> requiredUnavailableTime = getUnavailableTimes(requiredAttendees, events);
+        ArrayList<TimeRange> optionalUnavailableTime = getUnavailableTimes(optionalAttendees, events);
+
+        // If there are no meetings for the required attendees, return the entire day
+        if (requiredUnavailableTime.size() == 0) {
+            return entireDay;
+        }
+
+        // Combine and consolidate the unavailable time for both optional and required attendees
+        ArrayList   <TimeRange> combinedTime = new ArrayList   <TimeRange> ();
```
Comment (kyracq) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: nit: unnecessary spaces after "ArrayList"

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -270,5 +274,200 @@ public void notEnoughRoom() {
...
+      TimeRange.fromStartEnd(TIME_1030AM,TimeRange.END_OF_DAY, true));
+    Assert.assertEquals(expected, actual);
+}
+
+    @Test 
+    public void noTimeAvailable() {
+        // No mandatory attendees and two optional attendees with no gaps.
+        // should return nothing.
+        Collection<Event> events = Arrays.asList(
+          new Event("Event 1",TimeRange.fromStartDuration(DAY_START, DURATION_WHOLE_DAY), Arrays.asList(PERSON_A)),
```
Comment (kyracq) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:463]: For a whole day time range you can use TimeRange.WHOLE_DAY declared in TimeRange.java

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (otobin) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:91]: fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (otobin) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: fixed! thank you

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (otobin) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: fixed

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (otobin) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:463]: true! I forgot about that

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (otobin) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: fixed

**Pull request merged**